File: cpp/src/sbm.cpp
#include "frtb/sbm.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace frtb {

double scale_rho(double rho, const std::string& scenario, double high, double low) {
    if (scenario == "medium") return rho;
    if (scenario == "high") return std::min(high * rho, 1.0);
    if (scenario == "low") return low * rho;
    throw std::invalid_argument("scale_rho: unknown scenario '" + scenario + "'");
}

double bucket_kb(const std::vector<double>& ws,
                 const std::function<double(std::size_t, std::size_t)>& rho) {
    for (double w : ws)
        if (!std::isfinite(w))
            throw std::invalid_argument("bucket_kb: weighted sensitivities must be finite");
    double total = 0.0;
    for (double w : ws) total += w * w;
    const std::size_t n = ws.size();
    for (std::size_t k = 0; k < n; ++k)
        for (std::size_t l = 0; l < n; ++l)
            if (k != l) total += rho(k, l) * ws[k] * ws[l];
    // The max(0,.) guards against negative rounding of the quadratic form.
    return std::sqrt(std::max(0.0, total));
}

AggregateResult aggregate_buckets(
    const std::map<std::string, double>& kb, const std::map<std::string, double>& ws_sum,
    const std::function<double(const std::string&, const std::string&)>& gamma) {
    if (kb.size() != ws_sum.size())
        throw std::invalid_argument("aggregate_buckets: kb and ws_sum must cover the same buckets");
    std::vector<std::string> buckets;  // std::map keys are already sorted
    for (const auto& kv : kb) {
        if (!ws_sum.count(kv.first))
            throw std::invalid_argument(
                "aggregate_buckets: kb and ws_sum must cover the same buckets");
        buckets.push_back(kv.first);
    }

    auto inner = [&](const std::map<std::string, double>& sb) {
        double total = 0.0;
        for (const std::string& b : buckets) {
            double k = kb.at(b);
            total += k * k;
        }
        for (const std::string& b : buckets)
            for (const std::string& c : buckets)
                if (b != c) total += gamma(b, c) * sb.at(b) * sb.at(c);
        return total;
    };

    std::map<std::string, double> sb0 = ws_sum;
    double in0 = inner(sb0);
    if (in0 >= 0.0) return {std::sqrt(in0), false, sb0};
    // S_b fallback: clamp the bucket sums into [-K_b, K_b] and recompute once.
    std::map<std::string, double> sb1;
    for (const std::string& b : buckets)
        sb1[b] = std::max(std::min(ws_sum.at(b), kb.at(b)), -kb.at(b));
    double in1 = inner(sb1);
    return {std::sqrt(std::max(0.0, in1)), true, sb1};
}

RiskClassCharge delta_vega_charge(const BucketWs& bucket_ws, const IntraRhoFn& intra_rho,
                                  const GammaFn& gamma, const std::string& scenario,
                                  double scenario_high, double scenario_low) {
    std::map<std::string, double> kb;
    std::map<std::string, double> ws_sum;
    for (const auto& [b, factors] : bucket_ws) {
        // std::map iteration = lexicographic key order = Python's sorted().
        std::vector<std::string> keys;
        std::vector<double> ws;
        for (const auto& kv : factors) {
            keys.push_back(kv.first);
            ws.push_back(kv.second);
        }
        const std::string& bucket = b;
        auto rho_fn = [&](std::size_t i, std::size_t j) {
            return scale_rho(intra_rho(bucket, keys[i], keys[j]), scenario, scenario_high,
                             scenario_low);
        };
        kb[b] = bucket_kb(ws, rho_fn);
        double s = 0.0;
        for (double w : ws) s += w;
        ws_sum[b] = s;
    }

    if (kb.empty()) return {0.0, {}, false};

    auto gamma_fn = [&](const std::string& b, const std::string& c) {
        return scale_rho(gamma(b, c), scenario, scenario_high, scenario_low);
    };
    AggregateResult agg = aggregate_buckets(kb, ws_sum, gamma_fn);
    return {agg.charge, kb, agg.used_fallback};
}

double psi(double a, double b) { return (a < 0.0 && b < 0.0) ? 0.0 : 1.0; }

std::pair<double, double> curvature_bucket_kb(
    const std::vector<double>& cvr_up, const std::vector<double>& cvr_dn,
    const std::function<double(std::size_t, std::size_t)>& rho) {
    if (cvr_up.size() != cvr_dn.size())
        throw std::invalid_argument("curvature_bucket_kb: up/down CVR lists must match");

    auto side = [&](const std::vector<double>& cvr) {
        double total = 0.0;
        for (double c : cvr) {
            double m = std::max(c, 0.0);
            total += m * m;
        }
        const std::size_t n = cvr.size();
        for (std::size_t k = 0; k < n; ++k)
            for (std::size_t l = 0; l < n; ++l)
                if (k != l) total += rho(k, l) * cvr[k] * cvr[l] * psi(cvr[k], cvr[l]);
        return std::sqrt(std::max(0.0, total));
    };

    const double k_up = side(cvr_up);
    const double k_dn = side(cvr_dn);
    auto sum = [](const std::vector<double>& v) {
        double s = 0.0;
        for (double x : v) s += x;
        return s;
    };
    if (k_up >= k_dn) return {k_up, sum(cvr_up)};  // ties -> up side (pinned)
    return {k_dn, sum(cvr_dn)};
}

RiskClassCharge curvature_charge(const BucketCvr& bucket_cvr, const IntraRhoFn& intra_rho,
                                 const GammaFn& gamma, const std::string& scenario,
                                 const FactorKeys& factor_keys, double scenario_high,
                                 double scenario_low) {
    std::map<std::string, double> kb;
    std::map<std::string, double> sb;
    for (const auto& [b, updn] : bucket_cvr) {
        const std::vector<std::string>& keys = factor_keys.at(b);
        if (keys.size() != updn.first.size())
            throw std::invalid_argument("curvature_charge: factor keys mismatch in bucket '" + b +
                                        "'");
        const std::string& bucket = b;
        auto rho_fn = [&](std::size_t i, std::size_t j) {
            double r = scale_rho(intra_rho(bucket, keys[i], keys[j]), scenario, scenario_high,
                                 scenario_low);
            return r * r;  // curvature correlation = scenario-scaled delta rho squared
        };
        auto [k, s] = curvature_bucket_kb(updn.first, updn.second, rho_fn);
        kb[b] = k;
        sb[b] = s;
    }

    if (kb.empty()) return {0.0, {}, false};

    std::vector<std::string> buckets;
    for (const auto& kv : kb) buckets.push_back(kv.first);
    double total = 0.0;
    for (const std::string& b : buckets) {
        double k = kb.at(b);
        total += k * k;
    }
    for (const std::string& b : buckets)
        for (const std::string& c : buckets)
            if (b != c) {
                double g = scale_rho(gamma(b, c), scenario, scenario_high, scenario_low);
                total += (g * g) * sb.at(b) * sb.at(c) * psi(sb.at(b), sb.at(c));
            }
    return {std::sqrt(std::max(0.0, total)), kb, false};
}

}  // namespace frtb

File: rust/src/sbm.rs
//! Sensitivities-Based Method (SBM) aggregation.
//!
//! Formulas (FRTB structure, educational parameter set):
//!
//! ```text
//! Weighted sensitivity   WS_k = RW_k * s_k
//! Within-bucket          K_b  = sqrt(max(0, sum_k WS_k^2
//!                               + sum_{k != l} rho_kl WS_k WS_l))
//! Across buckets         Charge = sqrt(max(0, sum_b K_b^2
//!                               + sum_{b != c} gamma_bc S_b S_c))
//! ```
//!
//! with `S_b = sum_k WS_k`; if the argument of the outer sqrt is negative
//! the S_b FALLBACK applies: `S_b = max(min(sum_k WS_k, K_b), -K_b)` and
//! the aggregate is recomputed (the `max(0, .)` guard is kept as
//! belt-and-braces against negative rounding).
//!
//! Correlation scenarios: for every rho and gamma,
//! high `rho -> min(1.25*rho, 1)`, medium `rho`, low `0.75*rho` (pinned
//! simplification of Basel's `max(2*rho - 1, 0.75*rho)`). The risk-class
//! charges are computed under each scenario, summed across risk classes,
//! and the SBM capital is the MAX of the three scenario totals.
//!
//! Curvature: per curvature risk factor `k` with delta sensitivity `s_k`,
//!
//! ```text
//! CVR_k+ = -( V(x_k up)   - V - RW_k^curv * s_k )
//! CVR_k- = -( V(x_k down) - V + RW_k^curv * s_k )
//! K_b+/- = sqrt(max(0, sum_k max(CVR_k,0)^2
//!               + sum_{k!=l} rho_kl CVR_k CVR_l psi(CVR_k, CVR_l)))
//! psi(a, b) = 0 if a < 0 and b < 0 else 1
//! K_b    = max(K_b+, K_b-), S_b = CVR sum of the winning side (tie -> up)
//! ```
//!
//! Across buckets curvature uses `gamma^2`, `psi(S_b, S_c)` and
//! `max(0, .)` inside the sqrt; there is no S_b fallback for curvature.
//! Curvature correlations are the squares of the (scenario-scaled) delta
//! correlations (pinned simplification, documented in `API_SPEC.md`).

use std::collections::BTreeMap;

use crate::error::{invalid, Result};

/// Correlation scenario applied to every rho and gamma.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Scenario {
    /// `min(1.25 * rho, 1.0)` — capped at 1.
    High,
    /// `rho` unchanged.
    Medium,
    /// `0.75 * rho` (pinned simplification).
    Low,
}

/// The three scenarios in the pinned evaluation order (high, medium, low).
pub const SCENARIOS: [Scenario; 3] = [Scenario::High, Scenario::Medium, Scenario::Low];

impl Scenario {
    /// Parse a scenario name; errors on anything but high/medium/low.
    pub fn parse(name: &str) -> Result<Scenario> {
        match name {
            "high" => Ok(Scenario::High),
            "medium" => Ok(Scenario::Medium),
            "low" => Ok(Scenario::Low),
            other => invalid(format!("Scenario: unknown scenario '{other}'")),
        }
    }

    /// Canonical lowercase name.
    pub fn name(self) -> &'static str {
        match self {
            Scenario::High => "high",
            Scenario::Medium => "medium",
            Scenario::Low => "low",
        }
    }
}

/// Apply the correlation scenario scaler; the high scenario is capped at 1.
pub fn scale_rho(rho: f64, scenario: Scenario, high: f64, low: f64) -> f64 {
    match scenario {
        Scenario::Medium => rho,
        Scenario::High => (high * rho).min(1.0),
        Scenario::Low => low * rho,
    }
}

/// Within-bucket charge
/// `K_b = sqrt(max(0, sum WS^2 + sum_{k!=l} rho WS_k WS_l))`.
///
/// `rho(k, l)` supplies the pairwise correlation for `k != l`. The
/// `max(0, .)` guard protects against negative rounding of the quadratic
/// form.
pub fn bucket_kb<R>(ws: &[f64], rho: R) -> Result<f64>
where
    R: Fn(usize, usize) -> Result<f64>,
{
    for &w in ws {
        if !w.is_finite() {
            return invalid("bucket_kb: weighted sensitivities must be finite");
        }
    }
    let mut total = 0.0;
    for &w in ws {
        total += w * w;
    }
    let n = ws.len();
    for k in 0..n {
        for l in 0..n {
            if k != l {
                total += rho(k, l)? * ws[k] * ws[l];
            }
        }
    }
    Ok(total.max(0.0).sqrt())
}

/// Across-bucket aggregation output (with the S_b fallback bookkeeping).
#[derive(Debug, Clone, PartialEq)]
pub struct AggregateResult {
    /// The across-bucket charge.
    pub charge: f64,
    /// Whether the S_b fallback branch was taken.
    pub used_fallback: bool,
    /// The S_b actually used (post-fallback if triggered).
    pub sb: BTreeMap<String, f64>,
}

/// Across-bucket aggregation with the FRTB S_b fallback rule.
///
/// First tries `S_b = sum_k WS_k`. If
/// `sum_b K_b^2 + sum_{b!=c} gamma S_b S_c < 0`, recomputes once with
/// `S_b = max(min(sum_k WS_k, K_b), -K_b)`.
pub fn aggregate_buckets<G>(
    kb: &BTreeMap<String, f64>,
    ws_sum: &BTreeMap<String, f64>,
    gamma: G,
) -> Result<AggregateResult>
where
    G: Fn(&str, &str) -> Result<f64>,
{
    if kb.len() != ws_sum.len() || kb.keys().any(|b| !ws_sum.contains_key(b)) {
        return invalid("aggregate_buckets: kb and ws_sum must cover the same buckets");
    }
    let buckets: Vec<&String> = kb.keys().collect();

    let inner = |sb: &BTreeMap<String, f64>| -> Result<f64> {
        let mut total = 0.0;
        for b in &buckets {
            total += kb[*b] * kb[*b];
        }
        for b in &buckets {
            for c in &buckets {
                if b != c {
                    total += gamma(b, c)? * sb[*b] * sb[*c];
                }
            }
        }
        Ok(total)
    };

    let sb0: BTreeMap<String, f64> = ws_sum.clone();
    let inner0 = inner(&sb0)?;
    if inner0 >= 0.0 {
        return Ok(AggregateResult { charge: inner0.sqrt(), used_fallback: false, sb: sb0 });
    }
    let sb1: BTreeMap<String, f64> = kb
        .iter()
        .map(|(b, &k)| (b.clone(), ws_sum[b].min(k).max(-k)))
        .collect();
    let inner1 = inner(&sb1)?;
    Ok(AggregateResult { charge: inner1.max(0.0).sqrt(), used_fallback: true, sb: sb1 })
}

// --------------------------------------------------------------------------
// Delta / vega charges
// --------------------------------------------------------------------------

/// Per-scenario charge with per-bucket K_b detail (for reporting/golden).
#[derive(Debug, Clone, PartialEq)]
pub struct RiskClassCharge {
    /// The risk-class charge under the requested scenario.
    pub charge: f64,
    /// Per-bucket within-bucket charges.
    pub kb: BTreeMap<String, f64>,
    /// Whether the delta/vega S_b fallback was triggered.
    pub used_fallback: bool,
}

/// Generic delta or vega charge for one risk class under one scenario.
///
/// * `bucket_ws`: `{bucket -> {factor_key -> WS}}` (factor keys iterate in
///   sorted order, matching the reference).
/// * `intra_rho`: `(bucket, factor_k, factor_l) -> medium correlation`.
/// * `gamma`: `(bucket_b, bucket_c) -> medium cross-bucket gamma`.
pub fn delta_vega_charge<R, G>(
    bucket_ws: &BTreeMap<String, BTreeMap<String, f64>>,
    intra_rho: R,
    gamma: G,
    scenario: Scenario,
    scenario_high: f64,
    scenario_low: f64,
) -> Result<RiskClassCharge>
where
    R: Fn(&str, &str, &str) -> Result<f64>,
    G: Fn(&str, &str) -> Result<f64>,
{
    let mut kb = BTreeMap::new();
    let mut ws_sum = BTreeMap::new();
    for (b, factors) in bucket_ws {
        let keys: Vec<&String> = factors.keys().collect();
        let ws: Vec<f64> = keys.iter().map(|k| factors[*k]).collect();
        let rho_fn = |i: usize, j: usize| -> Result<f64> {
            Ok(scale_rho(intra_rho(b, keys[i], keys[j])?, scenario, scenario_high, scenario_low))
        };
        kb.insert(b.clone(), bucket_kb(&ws, rho_fn)?);
        let mut s = 0.0;
        for w in &ws {
            s += w;
        }
        ws_sum.insert(b.clone(), s);
    }

    if kb.is_empty() {
        return Ok(RiskClassCharge { charge: 0.0, kb, used_fallback: false });
    }

    let gamma_fn = |b: &str, c: &str| -> Result<f64> {
        Ok(scale_rho(gamma(b, c)?, scenario, scenario_high, scenario_low))
    };
    let agg = aggregate_buckets(&kb, &ws_sum, gamma_fn)?;
    Ok(RiskClassCharge { charge: agg.charge, kb, used_fallback: agg.used_fallback })
}

// --------------------------------------------------------------------------
// Curvature
// --------------------------------------------------------------------------

/// FRTB psi: 0 when both CVR terms are negative, else 1.
pub fn psi(a: f64, b: f64) -> f64 {
    if a < 0.0 && b < 0.0 {
        0.0
    } else {
        1.0
    }
}

/// Within-bucket curvature charge.
///
/// Returns `(K_b, S_b)` where `K_b = max(K_b+, K_b-)` with
/// `K_b+/- = sqrt(max(0, sum_k max(CVR_k,0)^2 + sum_{k!=l} rho_kl CVR_k
/// CVR_l psi(CVR_k, CVR_l)))` and `S_b` is the sum of CVRs on the winning
/// side (up on ties). `rho` must already be the CURVATURE correlation
/// (delta rho squared).
pub fn curvature_bucket_kb<R>(cvr_up: &[f64], cvr_dn: &[f64], rho: R) -> Result<(f64, f64)>
where
    R: Fn(usize, usize) -> Result<f64>,
{
    if cvr_up.len() != cvr_dn.len() {
        return invalid("curvature_bucket_kb: up/down CVR lists must match");
    }
    let side = |cvr: &[f64]| -> Result<f64> {
        let mut total = 0.0;
        for &c in cvr {
            let m = c.max(0.0);
            total += m * m;
        }
        let n = cvr.len();
        for k in 0..n {
            for l in 0..n {
                if k != l {
                    total += rho(k, l)? * cvr[k] * cvr[l] * psi(cvr[k], cvr[l]);
                }
            }
        }
        Ok(total.max(0.0).sqrt())
    };
    let k_up = side(cvr_up)?;
    let k_dn = side(cvr_dn)?;
    let sum = |cvr: &[f64]| {
        let mut s = 0.0;
        for &c in cvr {
            s += c;
        }
        s
    };
    if k_up >= k_dn {
        Ok((k_up, sum(cvr_up)))
    } else {
        Ok((k_dn, sum(cvr_dn)))
    }
}

/// Curvature charge for one risk class under one scenario.
///
/// * `bucket_cvr`: `{bucket -> (CVR+ list, CVR- list)}` aligned with
///   `factor_keys[bucket]`.
/// * `intra_rho` / `gamma` supply medium DELTA correlations; they are
///   scenario-scaled then SQUARED for curvature (pinned simplification).
pub fn curvature_charge<R, G>(
    bucket_cvr: &BTreeMap<String, (Vec<f64>, Vec<f64>)>,
    intra_rho: R,
    gamma: G,
    scenario: Scenario,
    factor_keys: &BTreeMap<String, Vec<String>>,
    scenario_high: f64,
    scenario_low: f64,
) -> Result<RiskClassCharge>
where
    R: Fn(&str, &str, &str) -> Result<f64>,
    G: Fn(&str, &str) -> Result<f64>,
{
    let mut kb = BTreeMap::new();
    let mut sb = BTreeMap::new();
    for (b, (up, dn)) in bucket_cvr {
        let keys = factor_keys
            .get(b)
            .ok_or_else(|| crate::error::FrtbError::Invalid(format!(
                "curvature_charge: factor keys mismatch in bucket '{b}'"
            )))?;
        if keys.len() != up.len() {
            return invalid(format!("curvature_charge: factor keys mismatch in bucket '{b}'"));
        }
        let rho_fn = |i: usize, j: usize| -> Result<f64> {
            let r =
                scale_rho(intra_rho(b, &keys[i], &keys[j])?, scenario, scenario_high, scenario_low);
            Ok(r * r)
        };
        let (k, s) = curvature_bucket_kb(up, dn, rho_fn)?;
        kb.insert(b.clone(), k);
        sb.insert(b.clone(), s);
    }

    if kb.is_empty() {
        return Ok(RiskClassCharge { charge: 0.0, kb, used_fallback: false });
    }

    let buckets: Vec<&String> = kb.keys().collect();
    let mut total = 0.0;
    for b in &buckets {
        total += kb[*b] * kb[*b];
    }
    for b in &buckets {
        for c in &buckets {
            if b != c {
                let g = scale_rho(gamma(b, c)?, scenario, scenario_high, scenario_low);
                total += (g * g) * sb[*b] * sb[*c] * psi(sb[*b], sb[*c]);
            }
        }
    }
    Ok(RiskClassCharge { charge: total.max(0.0).sqrt(), kb, used_fallback: false })
}
